src/libmongoc/src/mongoc/mongoc-topology-description.c

```diff
@@ -649,6 +654,57 @@ _mongoc_topology_description_validate_max_staleness (
    return true;
 }
 
+static bool
+_check_any_server_less_than_wire_version_13 (const void *sd_,
+                                             void *any_too_old_)
+{
+   const mongoc_server_description_t *sd = sd_;
+   bool *any_too_old = any_too_old_;
+   if (sd->max_wire_version < 13) {
+      *any_too_old = true;
+   }
+   return true;
+}
+
+/**
+ * @brief Calculate the read mode that we should be using, based on what was
+ * requested and what is available in the topology.
+ *
+ * Per the CRUD spec, if the requested read mode is *not* primary, and *any*
+ * server in the topology has a wire version <13, we must override the read
+ * mode preference with "primary." Wire version 13 indicates support on a
+ * secondary server for using aggregate pipelines that contain writing stages
+ * (i.e. '$out' and '$merge').
+ */
+static mongoc_read_mode_t
+_calc_effective_read_mode (const mongoc_topology_description_t *td,
```

Comment:
Per my earlier comment, this may need to calculate the effective read _preference_ (not just mode).